src/node_i18n.cc

```diff
@@ -64,7 +79,122 @@ bool InitializeICUDirectory(const char* icu_data_path) {
+                         size_t length) {
+  UErrorCode status = U_ZERO_ERROR;
+  uint32_t options = UIDNA_DEFAULT;
+  options |= UIDNA_NONTRANSITIONAL_TO_UNICODE;
+  UIDNA* uidna = uidna_openUTS46(options, &status);
+  UIDNAInfo info = UIDNA_INFO_INITIALIZER;
+
+  int32_t len = uidna_nameToUnicodeUTF8(uidna,
+                                        input, length,
+                                        **buf, buf->length(),
+                                        &info,
+                                        &status);
+
+  if (status == U_BUFFER_OVERFLOW_ERROR) {
+    status = U_ZERO_ERROR;
+    buf->AllocateSufficientStorage(len);
+    len = uidna_nameToUnicodeUTF8(uidna,
+                                  input, length,
+                                  **buf, buf->length(),
+                                  &info,
+                                  &status);
+  }
+
+  if (U_FAILURE(status))
+    len = -1;
+
+  uidna_close(uidna);
+  return len;
+}
+
+static int32_t ToASCII(MaybeStackBuffer<char>* buf,
+                       const char* input,
+                       size_t length) {
+  UErrorCode status = U_ZERO_ERROR;
+  uint32_t options = UIDNA_DEFAULT;
+  options |= UIDNA_NONTRANSITIONAL_TO_ASCII;
+  UIDNA* uidna = uidna_openUTS46(options, &status);
+  UIDNAInfo info = UIDNA_INFO_INITIALIZER;
+
+  int32_t len = uidna_nameToASCII_UTF8(uidna,
+                                       input, length,
+                                       **buf, buf->length(),
+                                       &info,
+                                       &status);
+
+  if (status == U_BUFFER_OVERFLOW_ERROR) {
+    status = U_ZERO_ERROR;
+    buf->AllocateSufficientStorage(len);
+    len = uidna_nameToASCII_UTF8(uidna,
+                                 input, length,
+                                 **buf, buf->length(),
+                                 &info,
+                                 &status);
+  }
+
+  if (U_FAILURE(status))
+    len = -1;
+
+  uidna_close(uidna);
+  return len;
+}
+
+static void ToUnicode(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  CHECK_GE(args.Length(), 1);
+  CHECK(args[0]->IsString());
+  Utf8Value val(env->isolate(), args[0]);
+  MaybeStackBuffer<char> buf;
+  int32_t len = ToUnicode(&buf, *val, val.length());
+
+  if (len < 0) {
+    env->ThrowError("Cannot convert name to unicode");
+    return;
+  }
+
+  args.GetReturnValue().Set(
+    String::NewFromUtf8(env->isolate(),
+    *buf,
+    v8::NewStringType::kNormal,
+    len).ToLocalChecked());
+}
+
+static void ToASCII(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  CHECK_GE(args.Length(), 1);
+  CHECK(args[0]->IsString());
+  Utf8Value val(env->isolate(), args[0]);
+  MaybeStackBuffer<char> buf;
+  int32_t len = ToASCII(&buf, *val, val.length());
+
+  if (len < 0) {
+    env->ThrowError("Cannot convert name to ASCII");
+    return;
```

Comment:
should be able to put these two on the same line. `git grep "return.*Throw"` shows that's how it's done in 260 other places.